forest_lib/forest/plot.py

```diff
@@ -28,235 +50,303 @@ def __init__(self, config, var, time):
+        """Plot RGBA images"""
+        if not self.visible:
+            return
+        x, y, dw, dh, image = self.render_image(self.current_config,
+                                                self.current_var,
+                                                self.current_time)
+
+        self.bokeh_img_ds.data.update({
+            'image': [image],
+            'x': [x],
+            'y': [y],
+            'dw': [dw],
+            'dh': [dh]
+        })
+        self.bokeh_figure.title.text = self.current_title
 
-    def get_data(self, var_name=None):
-        config_data = self.dataset[self.current_config]['data']
-        if var_name:
-            data_cube = config_data.get_data(var_name=var_name,
-                                 selected_time = self.current_time)
+    @lru_cache(maxsize=32)
+    def render_image(self,
+                     current_config,
+                     current_var,
+                     current_time):
+        """Plot RGBA images"""
+        if self.uses_mappable(current_var):
+            self.plot_funcs[current_var]()
+            # HACK: self._shape2d is populated by self.get_data()
+            ni, nj = self._shape2d
+            shape = (ni - 1, nj - 1)
+            image = rgba_from_mappable(self.main_plot, shape)
         else:
-            data_cube = config_data.get_data(var_name=self.current_var,
-                                             selected_time=self.current_time)
-        return data_cube
-
-    def update_precip(self):
-
-        '''Update function for precipitation plots, called by update_plot() when
-        precipitation is the selected plot type.
+            # Rasterize Figure instance instead of mappable
+            y_start, y_end, x_start, x_end = self.region_dict[self.current_region]
+            current_figsize = (8.0, 6.0)
+            aspect_ratio = (y_end - y_start) / (x_end - x_start)
+            self.current_figure.set_figwidth(current_figsize[0])
+            self.current_figure.set_figheight(
+                round(self.current_figure.get_figwidth() * aspect_ratio, 2)
+            )
+            self.current_axes.cla()
+            self.current_axes.set_position([0, 0, 1, 1])
+            self.current_axes.set_xlim((x_start, x_end))
+            self.current_axes.set_ylim((y_start, y_end))
+            self.current_axes.xaxis.set_visible(False)
+            self.current_axes.yaxis.set_visible(False)
+            self.plot_funcs[current_var]()
+            self.current_figure.canvas.draw()
+            image = forest.util.get_image_array_from_figure(self.current_figure)
+
+        # Smooth high resolution imagery
+        max_ni, max_nj = 800, 600
+        ni, nj, _ = image.shape
+        if (ni > max_ni) or (nj > max_nj):
+            image = smooth_image(image, (max_ni, max_nj))
+
+        # Plot bokeh image rgba
+        x, y, dw, dh = self.get_x_y_dw_dh()
+        return x, y, dw, dh, image
+
+    @staticmethod
+    def uses_mappable(variable):
```

Comment:
You are right, it should be `_plot_uses_figure()` and the logic reversed since when all plots have migrated away from Figure this method will go away